### ISSUE TITLE
[clang-tidy] False-positive result of check readability-redundant-parentheses with an iterator

### ISSUE THREAD

**Issue by pkl97** [llvm/llvm-project#192438]
This code triggers a false-positive result from clang-tidy check readability-redundant-parentheses (see https://godbolt.org/z/zYrfcxo69):
```c++
#include <vector>
#include <memory>
#include <iostream>

struct S
{
    int getValue() const { return 0; }
};

int main()
{
    std::vector<std::shared_ptr<S>> v;
    v.push_back(std::make_shared<S>());

    std::vector<std::shared_ptr<S>>::const_iterator it = v.cbegin();
    std::cout << (*it)->getValue() << std::endl; // <-- this line triggers the warning

    return 0;
}
```
The issue occurs on the main branch (upcoming clang 23). In clang 22 there was no such warning.
